main.cpp

```diff
@@ -0,0 +1,337 @@
+    matrix[15] = 1;
+}
+
+void applyTranslation(GLfloat x, GLfloat y, GLfloat z, GLfloat* matrix) {
+    matrix[12] = x;
+    matrix[13] = y;
+    matrix[14] = z;
+}
+
+void printMatrix(const GLfloat* matrix, const char* name) {
+    printf("%s:\n", name);
+    for (int i = 0; i < 4; ++i) {
+        for (int j = 0; j < 4; ++j) {
+            printf("%f ", matrix[i * 4 + j]);
+        }
+        printf("\n");
+    }
+}
+
+void printVertices(const GLfloat* vertices, int vertexCount) {
+    printf("Vertices:\n");
+    for (int i = 0; i < vertexCount; i += 3) {
+        printf("%f %f %f\n", vertices[i], vertices[i + 1], vertices[i + 2]);
+    }
+}
+
+int main() {
+    // NOTE: Initialize GLFW
+    if (!glfwInit()) {
+        fprintf(stderr, "Failed to initialize GLFW\n");
+        return -1;
+    }
+
+    // NOTE: Setup GLFW window properties
+    glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 3);
+    glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 3);
+    glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);
+
+    // NOTE: Create a windowed mode window and its OpenGL context
+    GLFWwindow* window = glfwCreateWindow(WINDOW_WIDTH, WINDOW_HEIGHT, "Math - Quaternion with OpenGL - ESGI", NULL, NULL);
+    if (!window) {
+        fprintf(stderr, "Failed to open GLFW window.\n");
+        glfwTerminate();
+        return -1;
+    }
+
+    // NOTE: Make the window's context current
+    glfwMakeContextCurrent(window);
+
+    // NOTE: Initialize GLEW
+    glewExperimental = GL_TRUE;
+    if (glewInit() != GLEW_OK) {
+        fprintf(stderr, "Failed to initialize GLEW\n");
+        return -1;
+    }
+
+    // NOTE: Build and compile shaders
+    GLuint vertexShader = createShader(GL_VERTEX_SHADER, vertexShaderSource);
+    GLuint fragmentShader = createShader(GL_FRAGMENT_SHADER, fragmentShaderSource);
+    GLuint shaderProgram = createProgram(vertexShader, fragmentShader);
+    glDeleteShader(vertexShader);
+    glDeleteShader(fragmentShader);
+
+    // NOTE: Setup cube VAO and VBO
+    GLuint VAO[2], VBO[2], EBO[2];
+    glGenVertexArrays(2, VAO);
+    glGenBuffers(2, VBO);
+    glGenBuffers(2, EBO);
+
+    for (int i = 0; i < 2; ++i) {
+        glBindVertexArray(VAO[i]);
+
+        glBindBuffer(GL_ARRAY_BUFFER, VBO[i]);
+        glBufferData(GL_ARRAY_BUFFER, sizeof(vertices), vertices, GL_STATIC_DRAW);
+
+        glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, EBO[i]);
+        glBufferData(GL_ELEMENT_ARRAY_BUFFER, sizeof(indices), indices, GL_STATIC_DRAW);
+
+        glVertexAttribPointer(0, 3, GL_FLOAT, GL_FALSE, 3 * sizeof(GLfloat), (GLvoid*)0);
+        glEnableVertexAttribArray(0);
+
+        glBindBuffer(GL_ARRAY_BUFFER, 0);
+        glBindVertexArray(0);
+    }
+
+    // NOTE: Define basic quaternions for rotations
+    Quaternion q1 = Quaternion(cos(M_PI / 8), sin(M_PI / 8), 0, 0); // NOTE: 45-degree rotation around x-axis
+    Quaternion q2 = Quaternion(cos(M_PI / 8), 0, sin(M_PI / 8), 0); // NOTE: 45-degree rotation around y-axis
```

Comment:
Maybe we could use Quaternion::eulerAngles() for these instead? It would be a waste not to use it here

Reply:
Done in the last commit